File: src/service_bus/ServiceBusSingleton.cc
#include "ServiceBusSingleton.h"

#include "Utils.h"

#define LOG_LEVEL INFO_LEVEL
#include "Logger.h"

using namespace service_bus;
using namespace commons;

bool ServiceBusSingleton::INITIALIZED = false;
shared_ptr<ServiceBus> ServiceBusSingleton::SERVICE_BUS = shared_ptr<ServiceBus>{};
mutex ServiceBusSingleton::API_MUTEX;

// -------------------------------------------------------------------------------------------------
// Public methods

void ServiceBusSingleton::init(const string& host_id,
                               const string& known_peer,
                               unsigned int port_lower,
                               unsigned int port_upper) {
    lock_guard<mutex> semaphore(API_MUTEX);
    if (INITIALIZED) {
        Utils::error(
            "ServiceBusSingleton already initialized. "
            "ServiceBusSingleton::init() should be called only once.");
    } else {
        ServiceBus::initialize_statics({}, port_lower, port_upper);
        SERVICE_BUS = shared_ptr<ServiceBus>(new ServiceBus(host_id, known_peer));
        INITIALIZED = true;
        Utils::sleep(500);
    }
}

shared_ptr<ServiceBus> ServiceBusSingleton::get_instance() {
    lock_guard<mutex> semaphore(API_MUTEX);
    if (!INITIALIZED) {
        Utils::error(
            "Uninitialized ServiceBusSingleton. ServiceBusSingleton::init() "
            "must be called before ServiceBusSingleton::get_instance()");
        return shared_ptr<ServiceBus>{};  // To avoid warnings
    } else {
        return SERVICE_BUS;
    }
}

File: src/3rd_party_slots/rust_metta_bus_client/src/service_bus_singleton.rs
use std::sync::OnceLock;

use crate::bus::PATTERN_MATCHING_QUERY;
use crate::service_bus::ServiceBus;
use crate::types::BoxError;

static INITIALIZED: OnceLock<bool> = OnceLock::new();
static SERVICE_BUS: OnceLock<ServiceBus> = OnceLock::new();

#[derive(Debug, Default)]
pub struct ServiceBusSingleton;

impl ServiceBusSingleton {
	pub fn init(
		host_id: String, known_peer: String, port_lower: u32, port_upper: u32,
	) -> Result<(), BoxError> {
		INITIALIZED.set(true).expect("ServiceBusSingleton already initialized!");
		let service_bus = ServiceBus::new(
			host_id,
			known_peer,
			vec![PATTERN_MATCHING_QUERY.to_string()],
			port_lower,
			port_upper,
		)?;
		SERVICE_BUS.set(service_bus).expect("ServiceBusSingleton already initialized!");
		Ok(())
	}

	pub fn get_instance() -> ServiceBus {
		INITIALIZED.get().expect("ServiceBusSingleton not initialized!");
		let service_bus = SERVICE_BUS.get().expect("ServiceBusSingleton not initialized!");
		service_bus.clone()
	}
}
